File: apps/api/src/omnia_api/schemas/task_board.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

TaskBoardStatus = Literal["backlog", "in_progress", "review", "done"]
TaskBoardAssignee = Literal["alexey", "alexey_jr", "artem", "roman"]
TaskBoardPriority = Literal["low", "medium", "high"]


class TaskBoardTaskCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    title: str = Field(min_length=1, max_length=160)
    description: str = Field(default="", max_length=2000)
    status: TaskBoardStatus = "backlog"
    assignee: TaskBoardAssignee
    priority: TaskBoardPriority = "medium"


class TaskBoardTaskUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    title: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=2000)
    status: TaskBoardStatus | None = None
    assignee: TaskBoardAssignee | None = None
    priority: TaskBoardPriority | None = None


class TaskBoardTaskPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    description: str
    status: TaskBoardStatus
    assignee: TaskBoardAssignee
    priority: TaskBoardPriority
    position: int
    created_at: datetime
    updated_at: datetime

File: apps/web/src/lib/api/task-board.ts
import { apiFetch, apiUrl, postBlob } from "./client";

export type TaskBoardStatus = "backlog" | "in_progress" | "review" | "done";
export type TaskBoardAssignee = "alexey" | "alexey_jr" | "artem" | "roman";
export type TaskBoardPriority = "low" | "medium" | "high";

export interface TaskBoardAttachment {
  id: string;
  filename: string;
  content_type: string;
  size: number;
  created_at: string;
}

export interface TaskBoardTask {
  id: string;
  title: string;
  description: string;
  status: TaskBoardStatus;
  assignee: TaskBoardAssignee;
  priority: TaskBoardPriority;
  position: number;
  attachments: TaskBoardAttachment[];
  created_at: string;
  updated_at: string;
}

export interface CreateTaskPayload {
  title: string;
  description?: string;
  status?: TaskBoardStatus;
  assignee: TaskBoardAssignee;
  priority?: TaskBoardPriority;
}

export type UpdateTaskPayload = Partial<CreateTaskPayload>;

export interface TaskBoardApi {
  listTasks(): Promise<TaskBoardTask[]>;
  createTask(payload: CreateTaskPayload): Promise<TaskBoardTask>;
  updateTask(id: string, payload: UpdateTaskPayload): Promise<TaskBoardTask>;
  deleteTask(id: string): Promise<void>;
  uploadAttachment(id: string, file: File): Promise<TaskBoardAttachment>;
  deleteAttachment(taskId: string, attachmentId: string): Promise<void>;
  attachmentDownloadUrl(taskId: string, attachmentId: string): string;
}

export const taskBoardApi: TaskBoardApi = {
  listTasks: () => apiFetch<TaskBoardTask[]>("/api/task-board/tasks"),
  createTask: (payload) =>
    apiFetch<TaskBoardTask>("/api/task-board/tasks", {
      method: "POST",
      json: payload,
    }),
  updateTask: (id, payload) =>
    apiFetch<TaskBoardTask>(`/api/task-board/tasks/${id}`, {
      method: "PATCH",
      json: payload,
    }),
  deleteTask: (id) =>
    apiFetch<void>(`/api/task-board/tasks/${id}`, { method: "DELETE" }),
  uploadAttachment: (id, file) =>
    postBlob<TaskBoardAttachment>(
      `/api/task-board/tasks/${id}/attachments`,
      file,
      {
        timeoutMs: 60_000,
        headers: { "X-File-Name": encodeURIComponent(file.name) },
      },
    ),
  deleteAttachment: (taskId, attachmentId) =>
    apiFetch<void>(
      `/api/task-board/tasks/${taskId}/attachments/${attachmentId}`,
      { method: "DELETE" },
    ),
  attachmentDownloadUrl: (taskId, attachmentId) =>
    apiUrl(`/api/task-board/tasks/${taskId}/attachments/${attachmentId}`),
};
